consensus/wal.go

```diff
@@ -72,6 +76,10 @@ type baseWAL struct {
 	group *auto.Group
 
 	enc *WALEncoder
+
+	flushTicker     *time.Ticker
+	flushInterval   time.Duration
+	periodicFlushFn func(*baseWAL) error // Allows for dependency injection.
```

Comment:
Is there another way to check if the WAL was flushed? Maybe we can simply read from it and ensure the msg is there?

Reply:
The ideal way I'd like to check if the WAL was flushed, if I have to avoid introducing test-related state into the data structures, would be to:

1. Create a fresh new WAL
2. Start the WAL with the periodic flush interval set to `walTestFlushInterval`
3. Generate some data in the WAL (5 blocks)
4. Ensure that nothing's been committed yet
5. Wait for `walTestFlushInterval` milliseconds to allow for the flush to happen
6. Ensure that the blocks we generated were committed

I can't find a simple way, given the existing code, to get step 4 right without introducing additional state somewhere. When I test for block height 4 immediately after the call to `consensus/wal_generator.go#WALGenerateNBlocks` (i.e. at step 4, by calling `wal.SearchForEndHeight`) it returns successfully, which is where I'd want it to fail.

Any suggestions on how I'd accomplish that with the existing code?

I've just committed the code I'd ideally like to use to test this, but it doesn't work:

```
--- FAIL: TestWALPeriodicSync (0.34s)
        Error Trace:    wal_test.go:171
        Error:      	An error is expected but got nil.
        Test:       	TestWALPeriodicSync
        Messages:   	expected to get an error for height 4
FAIL
coverage: 31.7% of statements
FAIL	github.com/tendermint/tendermint/consensus	1.189s
```